### PR TITLE
boards/sensebox: Add support for SenseBox board

### PR DESCRIPTION
**Pull request by leandrolanzieri**
### Contribution description
Support for [SenseBox board](https://www.sensebox.de/). This board is based on the Atmel SAMD21G18A microcontroller.
sensebox-top

#### Implemented and tested on the board:
* LEDs
* Digital Pins
* UART1 and UART2
* I2C interface
* XBEE1 and XBEE2 ports (SPI)

#### Note:
The kit comes with several XBEE shields. For this porting the SX127X shield is supposed to be connected on the XBEE1 port, and the mSD-Bee shield on the XBEE2 port.

For more information about XBEE ports please read **boards/sensebox/doc.txt** file.

### CODE REVIEW

File: boards/sensebox/include/board.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#define I2C_EN_PORT           PORT->Group[PB]
+#define I2C_EN_MASK           (1 << 11)
+#define I2C_EN_PIN            GPIO_PIN(PB, 11)
+#define I2C_EN_MODE           GPIO_OUT
+
+#define I2C_ENABLE             (I2C_EN_PORT.OUTSET.reg = I2C_EN_MASK)
+#define I2C_DISABLE            (I2C_EN_PORT.OUTCLR.reg = I2C_EN_MASK)
+/** @} */
+
+
```
Comment (jia200x) [inline boards/sensebox/include/board.h]: Only 1 space is required here

File: boards/sensebox/Makefile.include [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# define the cpu used by Adafruit Feather M0 boards
```
Comment (aabadie) [inline boards/sensebox/Makefile.include]: Adafruit based ? ;)

File: boards/sensebox/board.c [hunk 3]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C)  2017 Inria
```
Comment (aabadie) [inline boards/sensebox/board.c]: Not sure Inria is involved here, is it ?

File: boards/sensebox/doc.txt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * ### On XBEE ports
+ * XBEE1 and XBEE2 ports are enabled by default. To disable any of them use
+ * ```XBEEn_DISABLE```. To re-enable use ```XBEEn_ENABLE```.
+ *
+ * On the board there is a reset circuit for this ports that delays the reset
+ * signal from the enable signal for about 300ms. Because of that a delay has
+ * been introduced in the ```board_init``` function, so that any SPI device
+ * connected to the ports is reset before use.
+ *
+ * #### SX127X radio
```
Comment (aabadie) [inline boards/sensebox/doc.txt]: Is there an sx127x on the board ? I would rather use a RN2483 module like [this one](https://shop.sodaq.com/lorabee-rn2483-order-now.html). It can be plugged directly on the XBee port and then use the rn2xx3 driver of RIOT.

File: boards/sensebox/include/board.h [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+/**
+ * @name SX127X
```
Comment (aabadie) [inline boards/sensebox/include/board.h]: Do you have a link to the sx127x module you want to use on the Xbee port ? This sounds weird to me.

File: examples/lua_REPL/Makefile [hunk 6]
```diff
@@ -18,12 +18,12 @@ BOARD_INSUFFICIENT_MEMORY := bluepill calliope-mini cc2650-launchpad \
...
-                             openmote-cc2538 openmote-b  pba-d-01-kw2x \
+                             openmote-b openmote-cc2538 pba-d-01-kw2x \
                              remote-pa remote-reva remote-revb samd21-xpro \
                              saml21-xpro samr21-xpro seeeduino_arch-pro \
-                             slstk3401a sltb001a slwstk6220a sodaq-autonomo \
-                             sodaq-explorer sodaq-one stk3600  stm32f3discovery \
-                             yunjia-nrf51822
+                             sensebox slstk3401a sltb001a slwstk6220a \
+                             sodaq-autonomo sodaq-explorer  stk3600 \
+                             stm32f3discovery yunjia-nrf51822
```
Comment (miri64) [inline examples/lua_REPL/Makefile]: Please keep the reference to `sodaq-one` here ;-)

File: examples/lua_REPL/Makefile [hunk 7]
```diff
@@ -18,12 +18,12 @@ BOARD_INSUFFICIENT_MEMORY := bluepill calliope-mini cc2650-launchpad \
                              b-l072z-lrwan1  cc2538dk ek-lm4f120xl  feather-m0 \
                              ikea-tradfri limifrog-v1 mbed_lpc1768  nrf6310 \
                              nucleo-f091rc  nucleo-l073rz  nz32-sc151 \
-                             openmote-cc2538 openmote-b  pba-d-01-kw2x \
+                             openmote-b openmote-cc2538 pba-d-01-kw2x \
```
Comment (miri64) [inline examples/lua_REPL/Makefile]: These changes are unrelated. If you want to sort them, please put that in a separate PR.

File: tests/unittests/Makefile [hunk 8]
```diff
@@ -48,8 +48,8 @@ BOARD_INSUFFICIENT_MEMORY := airfy-beacon \
                              nz32-sc151 \
                              opencm904 \
                              openmote \
-                             openmote-cc2538 \
                              openmote-b \
+                             openmote-cc2538 \
```
Comment (miri64) [inline tests/unittests/Makefile]: These changes are unrelated. If you want to sort them, please put that in a separate PR.

File: boards/sensebox/include/board.h (hunk 1 above)
Comment (leandrolanzieri) [reply boards/sensebox/include/board.h]: @jia200x Changed, thanks!

File: examples/lua_REPL/Makefile (hunk 7 above)
Comment (leandrolanzieri) [reply examples/lua_REPL/Makefile]: @miri64 Back to original order, thanks!

File: boards/sensebox/board.c [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * @ingroup     boards_sensebox
+ * @{
+ *
+ * @file
+ * @brief       Board specific implementations for the SenseBox boards.
+ *
+ * @author      Leandro Lanzieri <leandro.lanzieri@haw-hamburg.de>
+ *
```
Comment (jia200x) [inline boards/sensebox/board.c]: Extra linespace here

File: boards/sensebox/include/periph_conf.h [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+static const i2c_conf_t i2c_config[] = {
+    {
+        .dev      = &(SERCOM0->I2CM),
+        .speed    = I2C_SPEED_NORMAL,
+        .scl_pin  = GPIO_PIN(PA, 8),
+        .sda_pin  = GPIO_PIN(PA, 9),
+        .mux      = GPIO_MUX_C,
+        .gclk_src = GCLK_CLKCTRL_GEN_GCLK0,
+        .flags    = I2C_FLAG_NONE
+     },
```
Comment (jia200x) [inline boards/sensebox/include/periph_conf.h]: extra whitespace here

File: boards/sensebox/include/board.h [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+#define SX127X_PARAM_SPI_NSS                XBEE1_CS_PIN        /* D23 */
+
+#define SX127X_PARAM_RESET                  GPIO_UNDEF
+
+#define SX127X_PARAM_DIOMULTI               XBEE1_INT_PIN       /* D24 */
+
+#define SX127X_PARAM_PASELECT               (SX127X_PA_RFO)
+
+#define SX127X_PARAMS                     { .spi       = SX127X_PARAM_SPI,     \
```
Comment (jia200x) [inline boards/sensebox/include/board.h]: alignment here

File: boards/sensebox/Makefile.features [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# Put defined MCU peripherals here (in alphabetical order)
+#FEATURES_PROVIDED += periph_adc
```
Comment (jia200x) [inline boards/sensebox/Makefile.features]: Please adapt as required

File: boards/sensebox_samd21/board.c [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    XBEE2_ENABLE;
+
+    /* initialize I2C bus enable pin */
+    gpio_init(I2C_EN_PIN, I2C_EN_MODE);
+
+    /* enable I2C bus */
+    I2C_ENABLE;
+
+    /* delay for SPI reset. Read README.md for more information. */
+    for(uint32_t i = 0; i < 3000000; i++){
```
Comment (smlng) [inline boards/sensebox_samd21/board.c]: coding style space between `for (` and `) {`. Also I'd suggest to introduce a macro like `#define WAIT_FOR_SPI_RESET  (3000000LU)` for the loop and then use a simple `unsigned i` for the loop counter, the platform is 32Bit so should be safe.

File: boards/sensebox_samd21/include/board.h [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+#define I2C_EN_PIN            GPIO_PIN(PB, 11)
+#define I2C_EN_MODE           GPIO_OUT
+
+#define I2C_ENABLE            (I2C_EN_PORT.OUTSET.reg = I2C_EN_MASK)
+#define I2C_DISABLE           (I2C_EN_PORT.OUTCLR.reg = I2C_EN_MASK)
+/** @} */
+
+/**
+ * @name SX127X
+ * @brief SX127X configuration (on XBEE1 port). This particular board has
```
Comment (smlng) [inline boards/sensebox_samd21/include/board.h]: IMHO the `@brief` is not needed here, but rather a short description for/after `@name`

File: boards/sensebox_samd21/include/board.h [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+                                            .reset_pin = SX127X_PARAM_RESET,   \
+                                            .dio0_pin  = SX127X_PARAM_DIO0,    \
+                                            .dio1_pin  = SX127X_PARAM_DIO1,    \
+                                            .dio2_pin  = SX127X_PARAM_DIO2,    \
+                                            .dio3_pin  = SX127X_PARAM_DIO3,    \
+                                            .paselect  = SX127X_PARAM_PASELECT }
+/** @} */
+
+/**
+ * @name    HDC1000 Temperature & Humidity sensor
```
Comment (smlng) [inline boards/sensebox_samd21/include/board.h]: could use `@brief` here, as grouping is not required for a single `#define`, applies below for the TSL, too

File: boards/sensebox_samd21/board.c (hunk 13 above)
Comment (leandrolanzieri) [reply boards/sensebox_samd21/board.c]: Done, thanks! :smile:

**Comment by smlng** [comment]
mine, too

File: boards/sensebox/Makefile.features (hunk 12 above)
Comment (leandrolanzieri) [reply boards/sensebox/Makefile.features]: Done!

**Comment by jia200x** [comment]
ACK & GO
